File: project-simple/src/engine/recommender.js
// src/engine/recommender.js

import {
  hhmm_to_min,
  min_to_hhmm,
  is_open_with_margin,
  is_open_on_day_perhari,
  get_close_hhmm_for_day
} from './eta.js';

import {
  sssp_from_origin,
  build_route_to_poi
} from './routing.js';

const _cache = {};
const _CACHE_TTL = 900 * 1000; // 15 mins

export function recommend(
  origin_stop_id,
  origin_walk_min,
  depart_hhmm,
  weekday,
  filters,
  db,
  limit = 15
) {
  const depart_hour = parseInt(depart_hhmm.split(":")[0], 10);
  const depart_min_of_day = hhmm_to_min(depart_hhmm);

  // Cache key
  const types_tuple = (filters.types || []).slice().sort().join(",");
  const cache_key = [
    origin_stop_id,
    depart_hour,
    _day_type(weekday),
    types_tuple,
    filters.max_transfers !== undefined ? filters.max_transfers : 99,
    filters.max_eta_min !== undefined ? filters.max_eta_min : 90,
    filters.min_stay_hours !== undefined ? filters.min_stay_hours : 2,
  ].join("|");

  const now = Date.now();
  if (_cache[cache_key] && (now - _cache[cache_key].ts < _CACHE_TTL)) {
    return _cache[cache_key].data;
  }

  // SSSP
  const { dist, pred } = sssp_from_origin(
    origin_stop_id,
    origin_walk_min,
    depart_hour,
    db.route_to_stop_list,
    db.route_stop_pos,
    db.stop_to_route_dirs,
    db.eta_exact,
    db.route_avg_eta,
    db.wait_lookup
  );

  const candidates = [];
  const max_eta = parseFloat(filters.max_eta_min !== undefined ? filters.max_eta_min : 90);
  const max_transfers = parseInt(filters.max_transfers !== undefined ? filters.max_transfers : 99, 10);
  const min_stay = parseInt(filters.min_stay_hours !== undefined ? filters.min_stay_hours : 2, 10) * 60;
  const type_filter = (filters.types || []).map(t => t.trim());

  for (const poi of db.poi_list) {
    if (type_filter.length > 0 && !type_filter.includes(poi.type)) continue;
    if (!poi.nearest_stop_id) continue;

    const route_result = build_route_to_poi(dist, pred, poi);
    if (!route_result) continue;

    const eta = route_result.eta_total_min;
    const transfers = route_result.transfers;

    if (eta > max_eta) continue;
    if (transfers > max_transfers) continue;

    const arrive_hhmm = min_to_hhmm(depart_min_of_day + eta);
    const close_today = get_close_hhmm_for_day(poi, weekday);
    const [open_ok, remaining] = is_open_with_margin(arrive_hhmm, close_today, min_stay);
    
    if (!open_ok) continue;
    if (!is_open_on_day_perhari(poi, weekday)) continue;

    const score = _score(poi, eta, transfers, remaining);

    candidates.push({
      ..._poi_fields(poi),
      eta_total_min: eta,
      transfers: transfers,
      arrive_hhmm: arrive_hhmm,
      remaining_open_min: remaining,
      open_margin_ok: open_ok,
      recommendation_score: Math.round(score * 100) / 100,
      route_legs: _enrich_legs(route_result.route_legs, db.stops_by_id),
    });
  }

  candidates.sort((a, b) => b.recommendation_score - a.recommendation_score);
  
  const result = candidates.slice(0, limit).map((c, i) => ({
    rank: i + 1,
    ...c
  }));

  _cache[cache_key] = { data: result, ts: now };
  return result;
}

function _score(poi, eta, transfers, remaining_min) {
  const norm_inv = (val, lo = 0, hi = 90) => Math.max(0.0, 1.0 - (val - lo) / Math.max(hi - lo, 1));
  
  const eta_score = norm_inv(eta, 0, 90);
  const walk_dist_m = parseFloat(poi.walk_dist_m || 1200);
  const walk_score = norm_inv(walk_dist_m, 0, 1200);
  const transfer_score = Math.max(0.0, 1.0 - transfers / 4);
  const margin_score = Math.min(1.0, remaining_min / 480);
  const rating_score = (parseFloat(poi.rating || 3) - 1) / 4;
  const pop_score = Math.min(1.0, Math.log1p(parseFloat(poi.vote_count || 0)) / 10);

  return (
    0.35 * eta_score +
    0.20 * walk_score +
    0.20 * transfer_score +
    0.10 * margin_score +
    0.10 * rating_score +
    0.05 * pop_score
  ) * 100;
}

function _poi_fields(poi) {
  return {
    poi_id: poi.poi_id,
    name: poi.name,
    type: poi.type,
    lat: poi.lat,
    lon: poi.lon,
    rating: poi.rating,
    vote_count: poi.vote_count,
    open_hhmm: poi.open_hhmm,
    close_hhmm: poi.close_hhmm,
    open_days: poi.open_days,
    needs_review: !!parseInt(poi.needs_review || 0, 10),
    hours_source_type: poi.hours_source_type,
    nearest_stop_id: poi.nearest_stop_id,
    nearest_stop_name: poi.nearest_stop_name,
    walk_dist_m: poi.walk_dist_m,
    walk_time_min: poi.walk_time_min,
    walk_access_class: poi.walk_access_class,
    description: poi.description,
    htm_weekday: poi.htm_weekday,
    htm_weekend: poi.htm_weekend,
    image: poi.image,
  };
}

function _enrich_legs(legs, stops_by_id) {
  return legs.map(leg => {
    const l = { ...leg };
    const keys = ["stop_id", "board_stop_id", "alight_stop_id", "at_stop_id", "from_stop_id", "to_stop_id"];
    for (const key of keys) {
      const sid = l[key];
      if (sid && stops_by_id[sid]) {
        l[key.replace("_id", "_name")] = stops_by_id[sid].name || sid;
      }
    }
    return l;
  });
}

function _day_type(weekday) {
  const weekends = ["Sabtu", "Minggu", "Saturday", "Sunday"];
  return weekends.includes(weekday) ? "weekend" : "weekday";
}

// Added targeted routing endpoint
export function routeTo(
  origin_stop_id,
  origin_walk_min,
  depart_hhmm,
  dest_poi_id,
  db
) {
  const depart_hour = parseInt(depart_hhmm.split(":")[0], 10);
  
  const { dist, pred } = sssp_from_origin(
    origin_stop_id,
    origin_walk_min,
    depart_hour,
    db.route_to_stop_list,
    db.route_stop_pos,
    db.stop_to_route_dirs,
    db.eta_exact,
    db.route_avg_eta,
    db.wait_lookup
  );

  const poi = db.poi_by_id[dest_poi_id];
  if (!poi) return { found: false, message: "POI tidak ditemukan" };

  const route_result = build_route_to_poi(dist, pred, poi);
  if (!route_result) return { found: false, message: "Rute ke POI ini tidak tersedia" };

  const depart_min_of_day = hhmm_to_min(depart_hhmm);
  const eta = route_result.eta_total_min;
  const arrive_hhmm = min_to_hhmm(depart_min_of_day + eta);

  return {
    found: true,
    eta_total_min: eta,
    transfers: route_result.transfers,
    arrive_hhmm: arrive_hhmm,
    route_legs: _enrich_legs(route_result.route_legs, db.stops_by_id)
  };
}

File: deploy_backend/engine/recommender.py
"""
engine/recommender.py — weighted scoring + LRU cache.

Scoring weights (sesuai proposal H.5):
  ETA           0.35  (lebih cepat = lebih baik)
  walk_dist     0.20  (halte lebih dekat = lebih baik)
  transfers     0.20  (direct > 1 transfer > ...)
  open_margin   0.10  (sisa jam buka lebih banyak = lebih baik)
  rating        0.10
  popularity    0.05
  [v4] needs_review penalty dihapus — POI needs_review=1 sudah
       dikecualikan dari pool oleh gate di startup.py
"""

from __future__ import annotations
import math
import time
from typing import Optional

from .eta import (
    hhmm_to_min, min_to_hhmm,
    is_open_with_margin, is_open_on_day,
    is_open_on_day_perhari, get_close_hhmm_for_day,  # [v4]
)
from .routing import sssp_from_origin, build_route_to_poi

# ── simple TTL cache ───────────────────────────────────────────────────────
_cache: dict[tuple, dict] = {}
_CACHE_TTL = 900  # 15 menit

MAX_WALK_M = 1200.0
MIN_OPEN_MARGIN = 120  # 2 jam


def recommend(
    origin_stop_id: str,
    origin_walk_min: float,
    depart_hhmm: str,
    weekday: str,
    filters: dict,
    # data
    poi_list: list,
    stops_by_id: dict,
    route_to_stop_list: dict,
    route_stop_pos: dict,
    stop_to_route_dirs: dict,
    eta_exact: dict,
    route_avg_eta: dict,
    wait_lookup: dict,
    limit: int = 15,
) -> list[dict]:
    """
    1 SSSP call → score all 116 POI → return top-limit.
    """
    depart_hour = int(depart_hhmm.split(":")[0])
    depart_min_of_day = hhmm_to_min(depart_hhmm)

    # ── cache key ─────────────────────────────────────────────────────────
    types_tuple = tuple(sorted(filters.get("types", []) or []))
    cache_key = (
        origin_stop_id, depart_hour,
        _day_type(weekday), types_tuple,
        filters.get("max_transfers", 99),
        filters.get("max_eta_min", 90),
        filters.get("min_stay_hours", 2),
    )
    now = time.time()
    if cache_key in _cache and now - _cache[cache_key]["ts"] < _CACHE_TTL:
        return _cache[cache_key]["data"]

    # ── 1 Dijkstra SSSP ───────────────────────────────────────────────────
    dist, pred = sssp_from_origin(
        origin_stop_id=origin_stop_id,
        origin_walk_min=origin_walk_min,
        depart_hour=depart_hour,
        route_to_stop_list=route_to_stop_list,
        route_stop_pos=route_stop_pos,
        stop_to_route_dirs=stop_to_route_dirs,
        eta_exact=eta_exact,
        route_avg_eta=route_avg_eta,
        wait_lookup=wait_lookup,
    )

    # ── score each POI ────────────────────────────────────────────────────
    candidates = []
    max_eta = float(filters.get("max_eta_min", 90))
    max_transfers = int(filters.get("max_transfers", 99))
    min_stay = int(filters.get("min_stay_hours", 2)) * 60
    type_filter = [t.strip() for t in (filters.get("types") or [])]

    for poi in poi_list:
        # type filter
        if type_filter and poi.get("type") not in type_filter:
            continue

        # coverage filter
        if not poi.get("nearest_stop_id"):
            continue

        route_result = build_route_to_poi(dist, pred, poi)
        if route_result is None:
            continue

        eta = route_result["eta_total_min"]
        transfers = route_result["transfers"]

        # ETA filter
        if eta > max_eta:
            continue
        if transfers > max_transfers:
            continue

        # jam buka filter
        arrive_hhmm = min_to_hhmm(depart_min_of_day + eta)
        # [v4] gunakan jadwal per hari jika tersedia
        close_today = get_close_hhmm_for_day(poi, weekday)
        open_ok, remaining = is_open_with_margin(
            arrive_hhmm,
            close_today,
            min_margin_min=min_stay,
        )
        if not open_ok:
            continue
        if not is_open_on_day_perhari(poi, weekday):   # [v4] per-day check
            continue

        score = _score(poi, eta, transfers, remaining)

        candidates.append({
            **_poi_fields(poi),
            "eta_total_min": eta,
            "transfers": transfers,
            "arrive_hhmm": arrive_hhmm,
            "remaining_open_min": remaining,
            "open_margin_ok": open_ok,
            "recommendation_score": round(score, 2),
            "route_legs": _enrich_legs(route_result["route_legs"], stops_by_id),
        })

    candidates.sort(key=lambda x: -x["recommendation_score"])
    result = [{"rank": i + 1, **c} for i, c in enumerate(candidates[:limit])]

    _cache[cache_key] = {"data": result, "ts": now}
    return result


def _score(poi: dict, eta: float, transfers: int, remaining_min: int) -> float:
    # norm helpers (0→1)
    def norm_inv(val, lo=0, hi=90):
        return max(0.0, 1.0 - (val - lo) / max(hi - lo, 1))

    eta_score     = norm_inv(eta, 0, 90)
    walk_score    = norm_inv(float(poi.get("walk_dist_m", 1200)), 0, 1200)
    transfer_score = max(0.0, 1.0 - transfers / 4)
    margin_score  = min(1.0, remaining_min / 480)
    rating_score  = (float(poi.get("rating", 3)) - 1) / 4
    pop_score     = min(1.0, math.log1p(float(poi.get("vote_count", 0))) / 10)
    # [v4] needs_review_penalty dihapus — POI needs_review=1 sudah dikecualikan
    review_penalty = 0  # tidak diperlukan lagi

    return (
        0.35 * eta_score
        + 0.20 * walk_score
        + 0.20 * transfer_score
        + 0.10 * margin_score
        + 0.10 * rating_score
        + 0.05 * pop_score
        # - review_penalty  [v4] dihapus
    ) * 100


def _poi_fields(poi: dict) -> dict:
    return {
        "poi_id": poi.get("poi_id"),
        "name": poi.get("name"),
        "type": poi.get("type"),
        "lat": poi.get("lat"),
        "lon": poi.get("lon"),
        "rating": poi.get("rating"),
        "vote_count": poi.get("vote_count"),
        "open_hhmm": poi.get("open_hhmm"),
        "close_hhmm": poi.get("close_hhmm"),
        "open_days": poi.get("open_days"),
        "needs_review": bool(int(poi.get("needs_review", 0))),
        "hours_source_type": poi.get("hours_source_type"),
        "nearest_stop_id": poi.get("nearest_stop_id"),
        "nearest_stop_name": poi.get("nearest_stop_name"),
        "walk_dist_m": poi.get("walk_dist_m"),
        "walk_time_min": poi.get("walk_time_min"),
        "walk_access_class": poi.get("walk_access_class"),
        "description": poi.get("description"),
        "htm_weekday": poi.get("htm_weekday"),
        "htm_weekend": poi.get("htm_weekend"),
        "image": poi.get("image"),
    }


def _enrich_legs(legs: list, stops_by_id: dict) -> list:
    """Tambahkan stop_name ke setiap leg."""
    result = []
    for leg in legs:
        l = dict(leg)
        for key in ("stop_id", "board_stop_id", "alight_stop_id",
                    "at_stop_id", "from_stop_id", "to_stop_id"):
            sid = l.get(key)
            if sid and sid in stops_by_id:
                l[key.replace("_id", "_name")] = stops_by_id[sid].get("stop_name", stops_by_id[sid].get("name", sid))
        result.append(l)
    return result


def _day_type(weekday: str) -> str:
    weekends = {"Sabtu", "Minggu", "Saturday", "Sunday"}
    return "weekend" if weekday in weekends else "weekday"
